File: 0338-counting-bits/0338-counting-bits.c

int count_set_bits(int num)
{
    int ret_counter = 0;
    while(num > 0)
    {
        num &= num-1;
        ret_counter++;
    }
    return ret_counter;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* countBits(int n, int* returnSize) {
    int i=0, counter = 0;
    int* retArray = (int*) malloc(sizeof(int) * (n+1));
    *returnSize = n+1;
    while( n >= 0)
    {
        counter = count_set_bits(n);
        retArray[n] = counter;
        n--;
    }
    return retArray;
}

File: 0338-counting-bits/0338-counting-bits.java
class Solution {

    private int countSetBits(int n){
        int count = 0;
        while(n > 0){
            n &= (n-1);
            count++;
        }
        return count;
    }
    public int[] countBits(int n) {
        int[] retArray = new int[n + 1];
        
        while(n >= 0)
        {
            int counter = countSetBits(n);
            retArray[n] = counter;
            n--;
        }
        return retArray;
    }
}